### ISSUE TITLE
nix-shell escaping broken

### ISSUE THREAD

**Issue by Profpatsch** [NixOS/nix#644]
(current `unstable` branch of NixOS)

In each nix-shell I start, the escaping is broken:

```
\[\][nix-shell:~/code/openlab/labapi]$\[\]  
```

Arrow up and down consequently give:

```
 ^[[A^[[B
```

and backspacing destroys the cursor position.

**Comment by lethalman** [comment]
Are you sure you aren't already in a nix-shell?

**Comment by Profpatsch** [comment]
Yes, from normal shell.
I created the minimal derivation,

```
with import <nixpkgs> {};

stdenv.mkDerivation {
    name = "foobar";
}
```

switched to `sh` and called `nix-shell`, and it still happens.

**Comment by Profpatsch** [comment]
Is it normal that starting a shell inside another shell breaks the arrow keys?

**Comment by Profpatsch** [comment]
I tried another terminal emulator, same result.

**Comment by Profpatsch** [comment]
Happens on tty, too.

**Comment by lethalman** [comment]
Yes because the first nix-shell uses the bash from your profile, which is interactive (with readline). The nix-shell however inside will probably have a non-interactive bash from stdenv. Hence the second nix-shell will spawn the non-interactive bash taken from the PATH.

**Comment by Profpatsch** [comment]
I’m not spawning to nix-shell inside each other, though.

**Comment by lethalman** [comment]
Then you have a non-interactive bash in PATH somewhere.

**Comment by Profpatsch** [comment]
htop shows the following:

```
lilyterm
|- /run/current-system/sw/bin/bash
  |- bash --rc-file /run/user/1000/nix-shell.x9URzx/rc

**Comment by wmertens** [comment]
Check that you didn't write bash anywhere in your config, you should write
bashInteractive there.

On Mon, Sep 14, 2015, 12:46 Profpatsch <notifications@github.com> wrote:

> htop shows the following:
>
> lilyterm
> |- /run/current-system/sw/bin/bash
>   |- bash --rc-file /run/user/1000/nix-shell.x9URzx/rc
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/NixOS/nix/issues/644#issuecomment-140035279>.
>
-- 

Wout.
(typed on mobile, excuse terseness)

**Comment by Profpatsch** [comment]
@wmertens You mean `users.extraUsers.user.shell` should not be `bash`?

What’s up with the file missing?

**Comment by wmertens** [comment]
Yes it should be bashInteractive or just the system shell

File missing?

On Mon, Sep 14, 2015, 17:08 Profpatsch <notifications@github.com> wrote:

> @wmertens <https://github.com/wmertens> You mean
> users.extraUsers.user.shell should not be bash?
>
> What’s up with the file missing?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/NixOS/nix/issues/644#issuecomment-140111133>.
>
-- 

Wout.
(typed on mobile, excuse terseness)

**Comment by Profpatsch** [comment]
See my comment above (https://github.com/NixOS/nix/issues/644#issuecomment-140035279). The bash is called with an rc file that doesn’t even exist.

**Comment by pmahoney** [comment]
I have this same problem in darwin (Yosemite) and Nix 1.10 (though I've had it with older versions of same as well):

```
$ cat shell.nix
with import <nixpkgs> {};

stdenv.mkDerivation {
    name = "foobar";
}

$ lsof -p $$ | grep -E '(^COMMAND|0u)'
COMMAND   PID     USER   FD   TYPE DEVICE  SIZE/OFF   NODE NAME
bash    72041 pmahoney    0u   CHR   16,3 0t2338609   1611 /dev/ttys003

$ cat ~/.bash_profile ~/.bashrc
cat: /Users/pmahoney/.bash_profile: No such file or directory
cat: /Users/pmahoney/.bashrc: No such file or directory

$ nix-shell
\[\][nix-shell:~]$\[\]

\[\][nix-shell:~]$\[\] ^P^A^N     ## <- result of typing ctrl-p, ctrl-a, ctrl-n

\[\][nix-shell:~]$\[\] ruby -e 'puts $stdin.tty?'
true

\[\][nix-shell:~]$\[\] lsof -p $$ | grep -E '(^COMMAND|0u)'
COMMAND   PID     USER   FD   TYPE DEVICE  SIZE/OFF   NODE NAME
bash    77200 pmahoney    0u   CHR   16,3 0t2339555   1611 /dev/ttys003
```

File descriptor zero in both shells points to `/dev/ttys003`.

**Comment by pmahoney** [comment]
Ok, now I understand the previous comments about bashInteractive.

`nix-env -iA nixpkgs.bashInteractive` (and uninstalling plain `nixpkgs.bash`) does the trick.

**Comment by Profpatsch** [comment]
It doesn’t happen anymore on my system. Not sure what was causing it.

**Comment by zimbatm** [comment]
Still having the issue on NixOS, even after installing bashInteractive.

**Comment by Profpatsch** [comment]
@zimbatm 
```bash
echo $PAGER
export PAGER='less -R'
```

**Comment by zimbatm** [comment]
```
[zimbatm@xps13:~/.../zimbatm/food-and-nutrition-conference]$ echo $PS1
\[\e]0; \u@\h: \w\a\]\[\033[G\]\[\e[1;32m\]\n\[\033[1;32m\][\u@\h:\w]\$\[\033[0m\] \[\e[0m\]

[zimbatm@xps13:~/.../zimbatm/food-and-nutrition-conference]$ echo $PAGER
less -R

[zimbatm@xps13:~/.../zimbatm/food-and-nutrition-conference]$ nix-shell 

\[\][nix-shell:~]$\[\] ^[[A^[[A^C
```

**Comment by Profpatsch** [comment]
Oh, then it’s yet another problem. Is that your user shell? Have you set it up via `programs.<shell>.enable = true`?

**Comment by zimbatm** [comment]
I'm using the default bash shell and have `programs.bash.enableCompletion = true;`. The first directives in my `~/.bashrc` are:
```
if [[ -z "$PS1" ]]; then return; fi
if [[ -n $IN_NIX_SHELL ]]; then return; fi
```

**Comment by Profpatsch** [comment]
@zimbatm What’s the `shell` attribute of your user?

**Comment by zimbatm** [comment]
```nix
  # User accounts
  users.extraUsers.zimbatm = {
    isNormalUser = true;
    extraGroups = [
      "wheel"
      "audio" "sound" "video" "networkmanager" "input" "tty"
      "docker"
    ];
    uid = 1000;
    hashedPassword = "<snip>";
  };
```

**Comment by Profpatsch** [comment]
Have you tried commenting out `defaultUserShell`, adding `shell = "${pkgs.bash}/bin/bash` to your user, enabling `programs.bash`, (re)moving your `.bashrc` so that it can not be loaded and then rebooting?
If that doesn’t help we’ve got a more serious bug somewhere. If it does, you could try different combinations of the options and your bashrc respectively.

Oh, and I’ve got the following in my config:

```
    ########
    # Fixes

    # fix for emacs ssh
    programs.bash.promptInit = "PS1=\"# \"";
```

**Comment by mikefaille** [comment]
Hi, I'm not sure how I just get this bug too after one week on NixOS-Unstable as desktop.
To repeat this problem, I need to enter in my bash shell using gnome-terminal. And, typing `bash` to enter in new subshell. Or, if I use nix-shell, I get this problem too. When I get the error we saw in 1rst post, I can't auto-complete or even use backspace.

BTW, this line in my config fix the problem... : 
```
programs.bash.promptInit = "PS1=\"# \"";
```
Although, if I use this last line in my config, in new terminal, I can enter in subshell. But, in sub-terminal, I can't. Thank you if you can give me procedures to debug this.

**Comment by domenkozar** [comment]
@mikefaille did you try any of the above recommendations how to fix the problem?

**Comment by mikefaille** [comment]
@domenkozar  
I tried only one option at once : 
```
users.defaultUserShell = pkgs.bashInteractive
users.extraUsers.michael.shell = "${pkgs.bash}/bin/bash"
```
and it still dont work after going in new subshell.

Maybe the problem is in /etc/profile. My mine : 
https://gist.github.com/mikefaille/0b08085aafef1a925dd87ec26b92a8c2

Actually, `/etc/bashrc` is sourced after theses instructions : 
```bash
# Only execute this file once per shell.
if [ -n "$__ETC_PROFILE_SOURCED" ]; then return; fi
__ETC_PROFILE_SOURCED=1

# Prevent this file from being sourced by interactive non-login child shells.
export __ETC_PROFILE_DONE=1
```

**Comment by mikefaille** [comment]
Ah.... I just figured that I have 2 bash binaries installed by just tapping `whereis bash`.
After, I checked in my user env. for installed package using `nix-env -q --installed`; nothing. Then, I checked installed package as root `sudo nix-env -q --installed` and I saw bash installed in my root user env... I deinstall bash as root `sudo nix-env -e bash`. I tried to open new subshell and... it work ! 

And, I can reproduce my issue. I just need to install bash as user or root using 
```
nix-env -i bash
``` 
to get `\[\][michael@nixos:~]$\[\]` in subshell.

If I really want to install bash, `bash-interactive` package must be used.

**Comment by Mic92** [comment]
@zimbatm this also happens when using nix-shell in an direnv enabled directory (where nix was used).

**Comment by iclanzan** [comment]
I am having the same problem. I can see there are 2 bash binaries under `nix-shell` when running `whereis bash` but the root user doesn’t have a separate one and I removed all the `buildInputs` from `default.nix` so I am not sure where the second bash is coming from.

**Comment by Mic92** [comment]
@iclanzan bash is in your stdenv, just add `bashInteractive`.

**Comment by iclanzan** [comment]
Adding `bashInteractive` to `buildInputs` doesn’t help. I can see 3 binaries now but the wrong one is still being used because something is setting the `SHELL` env var to the non-interactive bash binary when running `nix-shell`.

I guess a workaround can be `nix-shell --command "export SHELL='path/to/good/bash'; return"`

**Comment by Mic92** [comment]
Does this one works?

```
with import <nixpkgs> {};
stdenv.mkDerivation {
  name = "env";
  SHELL="${bashInteractive}/bin/bash";
  # ...
}
```

**Comment by iclanzan** [comment]
Nope. `SHELL` is being overwritten. (I also set `SHELL1` to the same value to verify)

**Comment by unode** [comment]
I'm having a somewhat similar problem.
```
nix-shell -p hello
```
works as expected, whereas
```
nix-shell -p hello --pure
```
causes <kbd>BackSpace</kbd> to move the cursor forward 1 position instead of backwards:

[![asciicast](https://asciinema.org/a/SOTKagmNnczlpgpmnwxFOKpAK.png)](https://asciinema.org/a/SOTKagmNnczlpgpmnwxFOKpAK)

**Comment by tomberek** [comment]
Closing as stale. Re-open if needed.
